src/NHibernate/Type/CollectionType.cs

```diff
@@ -391,14 +392,36 @@ public virtual object InstantiateResult(object original)
 			// in which case we should never ever ever reset the dirty flag
 			// on the target because we simply do not know...
 			IPersistentCollection originalPc = original as IPersistentCollection;
-			IPersistentCollection resultPc = result as IPersistentCollection;
-			if(originalPc != null && resultPc!=null)
+			IPersistentCollection resultPc = target as IPersistentCollection;
+			if(resultPc!=null)
 			{
-				if (!originalPc.IsDirty)
+				if (elementsEqual || (originalPc != null && !originalPc.IsDirty))
```

Comment:
I do not see why this would be valid. Why original equality with target warrants target to be non-dirty, especially if original was dirty?  
Reading [NH-3932](https://nhibernate.jira.com/browse/NH-3932), this PR is about avoiding unneeded updates when the original is not a persistent collection. This is precisely the case the comment just above those lines writes we should never ever reset the dirty flag, cause we do not know... If this comment is just plain wrong, it should be amended. But I do not believe it to be wrong. It looks to me this change will optimize some cases will breaking other cases.

At the very least, condition for reseting the dirty flag should be hardened in my opinion. Maybe something like:

- check before clearing target if it was already flagged dirty (save a `targetDirty` variable, value irrelevant if target is not a persistent collection),
- if original is a `IPersistentCollection`:
  - if original is not dirty, reset dirty.
- **else** (original not a `IPersistentCollection`) if elements were equals and target was not dirty, reset dirty.

But even this, I am not sure it would be valid for all cases.